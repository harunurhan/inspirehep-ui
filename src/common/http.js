import axios from 'axios';

const http = axios.create({
  baseURL: '/api',
  headers: {
    Accept: 'application/vnd+inspire.record.ui+json',
  },
});

export default http;
