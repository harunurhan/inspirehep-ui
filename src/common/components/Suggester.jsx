import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { AutoComplete } from 'antd';
import http from '../http';

class Suggester extends Component {
  constructor(props) {
    super(props);

    this.onSearch = this.onSearch.bind(this);

    this.state = {
      results: [],
    };
  }

  async onSearch(value) {
    if (!value) {
      this.setState({ results: [] });
      return;
    }

    const { pidType, suggesterName } = this.props;
    const urlWithQuery = `/${pidType}/_suggest?${suggesterName}=${value}`;
    try {
      const response = await http.get(urlWithQuery);
      const results = this.responseDataToResults(response.data);
      this.setState({ results });
    } catch (error) {
      this.setState({ results: [] });
    }
  }

  responseDataToResults(responseData) {
    const { suggesterName } = this.props;
    return responseData[suggesterName][0].options;
  }

  render() {
    const { results } = this.state;
    const {
      renderResultItem,
      suggesterName,
      pidType,
      ...autoCompleteProps
    } = this.props;
    return (
      <AutoComplete onSearch={this.onSearch} {...autoCompleteProps}>
        {results.map(result => (
          <AutoComplete.Option key={result._id}>
            {renderResultItem(result)}
          </AutoComplete.Option>
        ))}
      </AutoComplete>
    );
  }
}

Suggester.propTypes = {
  pidType: PropTypes.string.isRequired,
  suggesterName: PropTypes.string.isRequired,
  renderResultItem: PropTypes.func,
};

Suggester.defaultProps = {
  renderResultItem: resultItem => resultItem.text,
};

export default Suggester;
