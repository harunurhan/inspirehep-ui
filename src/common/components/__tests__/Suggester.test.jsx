import React from 'react';
import { shallow } from 'enzyme';
import MockAdapter from 'axios-mock-adapter';

import http from '../../http';
import Suggester from '../Suggester';

const mockHttp = new MockAdapter(http);

describe('Suggester', () => {
  afterEach(() => {
    mockHttp.reset();
  });

  it('renders results onSearch', async () => {
    const suggesterQueryUrl = '/literature/_suggest?abstract_source=test';
    const reponseData = {
      abstract_source: [
        {
          options: [
            {
              _id: '1',
              text: 'Result 1',
            },
            {
              _id: '2',
              text: 'Result 2',
            },
          ],
        },
      ],
    };
    mockHttp.onGet(suggesterQueryUrl).replyOnce(200, reponseData);
    const wrapper = shallow(
      <Suggester pidType="literature" suggesterName="abstract_source" />
    );
    await wrapper.instance().onSearch('test');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders results with custom template onSearch', async () => {
    const suggesterQueryUrl = '/literature/_suggest?abstract_source=test';
    const reponseData = {
      abstract_source: [
        {
          options: [
            {
              _id: '1',
              text: 'Result 1',
              extra: 'Extra 1',
            },
            {
              _id: '2',
              text: 'Result 2',
              extra: 'Extra 2',
            },
          ],
        },
      ],
    };
    mockHttp.onGet(suggesterQueryUrl).replyOnce(200, reponseData);
    const wrapper = shallow(
      <Suggester
        pidType="literature"
        suggesterName="abstract_source"
        renderResultItem={result => `${result.text} - ${result.extra}`}
      />
    );
    await wrapper.instance().onSearch('test');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders empty if request fails', async () => {
    const suggesterQueryUrl = '/literature/_suggest?abstract_source=test';
    mockHttp.onGet(suggesterQueryUrl).replyOnce(404);
    const wrapper = shallow(
      <Suggester pidType="literature" suggesterName="abstract_source" />
    );
    await wrapper.instance().onSearch('test');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
