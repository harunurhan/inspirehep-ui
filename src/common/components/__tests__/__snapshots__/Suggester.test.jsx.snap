// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggester renders empty if request fails 1`] = `
<AutoComplete
  choiceTransitionName="zoom"
  filterOption={false}
  onSearch={[Function]}
  optionLabelProp="children"
  prefixCls="ant-select"
  showSearch={false}
  transitionName="slide-up"
/>
`;

exports[`Suggester renders results onSearch 1`] = `
<AutoComplete
  choiceTransitionName="zoom"
  filterOption={false}
  onSearch={[Function]}
  optionLabelProp="children"
  prefixCls="ant-select"
  showSearch={false}
  transitionName="slide-up"
>
  <Option
    key="1"
  >
    Result 1
  </Option>
  <Option
    key="2"
  >
    Result 2
  </Option>
</AutoComplete>
`;

exports[`Suggester renders results with custom template onSearch 1`] = `
<AutoComplete
  choiceTransitionName="zoom"
  filterOption={false}
  onSearch={[Function]}
  optionLabelProp="children"
  prefixCls="ant-select"
  showSearch={false}
  transitionName="slide-up"
>
  <Option
    key="1"
  >
    Result 1 - Extra 1
  </Option>
  <Option
    key="2"
  >
    Result 2 - Extra 2
  </Option>
</AutoComplete>
`;
